src/primitive/particle.rs

```diff
@@ -0,0 +1,272 @@
+        let ray = Ray::new(*start, (*end - *start).normalize());
+        match self.intersection(&ray) {
+            None => false,
+            Some(p) => (p - *start).magnitude2() <= (*end - *start).magnitude2(),
+        }
+    }
+}
+
+impl<P> Discrete<Ray<P::Scalar, P, P::Diff>> for Particle<P>
+where
+    P: EuclideanSpace,
+    P::Diff: InnerSpace,
+    P::Scalar: BaseFloat,
+{
+    /// Ray needs to be in particle object space
+    fn intersects(&self, ray: &Ray<P::Scalar, P, P::Diff>) -> bool {
+        P::from_value(P::Scalar::zero()).intersects(ray)
+    }
+}
+
+impl<P> Continuous<Ray<P::Scalar, P, P::Diff>> for Particle<P>
+where
+    P: EuclideanSpace,
+    P::Diff: InnerSpace,
+    P::Scalar: BaseFloat,
+{
+    type Result = P;
+
+    /// Ray needs to be in particle object space
+    fn intersection(&self, ray: &Ray<P::Scalar, P, P::Diff>) -> Option<P> {
+        P::from_value(P::Scalar::zero()).intersection(ray)
+    }
+}
+
+impl<P, C> DiscreteTransformed<(Particle<P>, P, P)> for C
+where
+    C: ContinuousTransformed<Ray<P::Scalar, P, P::Diff>, Result = P, Point = P>,
+    P: EuclideanSpace,
+    P::Diff: InnerSpace,
+    P::Scalar: BaseFloat,
+{
+    type Point = P;
+
+    fn intersects_transformed<T>(
+        &self,
+        &(_, ref start, ref end): &(Particle<P>, P, P),
+        transform: &T,
+    ) -> bool
+    where
+        T: Transform<P>,
+    {
+        let ray = Ray::new(*start, (*end - *start).normalize());
+        match self.intersection_transformed(&ray, transform) {
+            None => false,
+            Some(p) => (p - *start).magnitude2() <= (*end - *start).magnitude2(),
+        }
+    }
+}
+
+impl<P, C> Continuous<(Particle<P>, P, P)> for C
+where
+    C: Continuous<Ray<P::Scalar, P, P::Diff>, Result = P>,
+    P: EuclideanSpace,
+    P::Diff: InnerSpace,
+    P::Scalar: BaseFloat,
+{
+    type Result = P;
+
+    fn intersection(&self, &(_, ref start, ref end): &(Particle<P>, P, P)) -> Option<P> {
+        let ray = Ray::new(*start, (*end - *start).normalize());
+        match self.intersection(&ray) {
+            None => None,
+            Some(p) => {
+                if (p - *start).magnitude2() <= (*end - *start).magnitude2() {
+                    Some(p)
+                } else {
+                    None
+                }
+            }
+        }
+    }
+}
+
+impl<P, C> ContinuousTransformed<(Particle<P>, P, P)> for C
+where
+    C: ContinuousTransformed<Ray<P::Scalar, P, P::Diff>, Result = P, Point = P>,
+    P: EuclideanSpace,
+    P::Diff: InnerSpace,
+    P::Scalar: BaseFloat,
+{
+    type Point = P;
+    type Result = P;
+
+    fn intersection_transformed<T>(
+        &self,
+        &(_, ref start, ref end): &(Particle<P>, P, P),
+        transform: &T,
+    ) -> Option<P>
+    where
+        T: Transform<P>,
+    {
+        let ray = Ray::new(*start, (*end - *start).normalize());
+        match self.intersection_transformed(&ray, transform) {
```

Comment:
fyi, for this very pattern there is [and_then](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.and_then)

Reply:
Ah, thanks.